### ISSUE TITLE
hypertable_size() returns NULL for empty hypertable

### ISSUE THREAD

**Issue by RobAtticus** [timescale/timescaledb#2871]
**Relevant system information:**
 - TimescaleDB version (output of `\dx` in `psql`): at least 2.0.0

**Describe the bug**
When calling `hypertable_size` on a hypertable that has no rows, the result is `NULL` whereas one might reasonably expect it to return `0`.

**To Reproduce**
1. Create a hypertable, e.g., `foo`
2. Call `hypertable_size` on it before any rows are inserted:
```
SELECT * FROM hypertable_size('bar');
 hypertable_size
-----------------

(1 row)
```

**Expected behavior**
A single row with 0 for the size.

**Actual behavior**
A single row with NULL returned.

**Comment by NunoFilipeSantos** [comment]
Thank you @RobAtticus. Added to our triaging board for consideration.

**Comment by NunoFilipeSantos** [comment]
@erimatnor please add with versions of OS you are using and also the FS you have on both.

**Comment by erimatnor** [comment]
@NunoFilipeSantos OS versions no longer relevant. Figured out that testing issues for the PR were related to running the test in parallel with other tests rather than being OS-related.

In addition, PR review brought up the following:

* Should `hypertable_size()` and similar functions raise error when called on non-hypertables and indexes or 1. just return NULL, or 2. return the regular table size. The behavior of existing PG size functions might inform us here.
* `hypertable_detailed_size()` should probably also return size of access node objects in addition to data node objects. But what should be in the `node_name` field for the access node? Just NULL or a fixed string "access node". The latter has issues if you name a data node "access node".

**Comment by erimatnor** [comment]
As alluded to above, fixing the issue of returning NULL on an empty table raised a couple of other related issues:

1. `hypertable_size` fails with `ERROR:  query returned no rows` when executed on anything but a hypertable (e.g., regular table, index, etc.).
    To fix this, the following solutions are proposed when this function is executed on a non-hypertable relation:
     1. Fail with a better error, e.g.: `function only supported on hypertables`
     2. Return `NULL`  (e.g., for regular tables, indexes, views, etc.)
     3. Return the corresponding size of "other" relation (e.g., table size, index size, etc.), otherwise NULL (or error)
2. `hypertable_size` doesn't include the size of objects on the root hypertable (e.g., empty indexes) 
     * On distributed hypertables, this extends to the root tables on data nodes and access node
     * The fix is to include the size of empty indexes on the root table
3. `hypertable_detailed_size` has similar issues as `hypertable_size`, and in addition:
     * `hypertable_detalied_size` returns a row for each data node, but not one for the access node although it contains non-zero size objects. The fix is to return also a row for the access nodes with size information about the root hypertable. (This would also work for storing some data on the access node if we want to support that in the future.)
     * A follow-up question is what to return in the `node_name` field in the row for the access node. Should it be `NULL` or some hard-coded string (e.g., `access_node`). Potential issue with the latter approach is a conflict with a data node named `access_node`.

**Comment by k-rus** [comment]
> * `hypertable_size` fails with `ERROR:  query returned no rows` when executed on anything but a hypertable (e.g., regular table, index, etc.).
>   To fix this, the following solutions are proposed when this function is executed on a non-hypertable relation:
>   
>   1. Fail with a better error, e.g.: `function only supported on hypertables`
>   2. Return `NULL`  (e.g., for regular tables, indexes, views, etc.)
>   3. Return the corresponding size of "other" relation (e.g., table size, index size, etc.), otherwise NULL (or error)

Option iii. can mislead a user as the result doesn't correspond to the name of the function, i.e, imho, semantically incorrect.

To choose between option i and ii it is good to understand use cases and how the actual code will look like with each option. In general I see two use cases:
1. A user manually interactively calls the function.
2. A user writes a tool (e.g., forge explorer) which will call the function.

More use cases?

For the use case 1, the option i. might be preferable as it is inclusive for newbies.
The option ii. is easier to use in automation, since dealing with value is easier than with error. In the case of option i, a user will need a wrapper function like:
```sql
create function hypertabl_size_no_error(relation) ->
select hypertable_size(h) from timescaledb_information.hypertables h where (h.schema_name||'.'||h.table_name)::regclass = relation;
```
Right?

**Comment by erimatnor** [comment]
> 
> Option iii. can mislead a user as the result doesn't correspond to the name of the function, i.e, imho, semantically incorrect.
> 

Maybe some info on the behavior of standard PG functions can be informative here:

```
postgres=# select pg_table_size(89232);
 pg_table_size 
---------------
              
(1 row)

create table foo (time timestamptz, device int, temp float);
CREATE TABLE
postgres=# create index foo_time_idx on foo (time);
CREATE INDEX

postgres=# select pg_table_size('foo');
 pg_table_size 
---------------
             0
(1 row)

postgres=# select pg_table_size('foo_time_idx');
 pg_table_size 
---------------
          8192
(1 row)

postgres=# select pg_table_size('foo_view');
 pg_table_size 
---------------
             0
(1 row)

postgres=# select pg_indexes_size('foo');
 pg_indexes_size 
-----------------
            8192
(1 row)

postgres=# select pg_indexes_size('foo_time_idx');
 pg_indexes_size 
-----------------
               0
(1 row)

postgres=# select pg_indexes_size('foo_view');
 pg_indexes_size 
-----------------
               0
(1 row)
```

**Comment by k-rus** [comment]
> It might actually be convenient if you return the hypertable size for hypertables, but default to, e.g., `pg_table_size()` for other objects.

This is convenient in one use case, but will be confusing in another use case.



> SELECT hypertable_size(oid), relname FROM pg_class

Actually this use case doesn't have value, imho, since it clearly mixes different semantics. I expect that a user will write the following query instead:
```sql
SELECT hypertable_size(hypertable_name), hypertable_name FROM timescaledb_information.hypertables;
```

If a user wants to get sizes for hypertables and PG tables and we want to provide a single function, then the name should correspond to semantics, e.g., `ts_table_size` or `ts_relation_size`, but not `hypertable_size`.

**Comment by erimatnor** [comment]
> > It might actually be convenient if you return the hypertable size for hypertables, but default to, e.g., `pg_table_size()` for other objects.
> 
> This is convenient in one use case, but will be confusing in another use case.
> 
> > SELECT hypertable_size(oid), relname FROM pg_class
> 
> Actually this use case doesn't have value, imho, since it clearly mixes different semantics. 

We still need to decide on the behavior for this use case, which can be any of:

* Raise error
* Return NULL
* Return 0
* Return size of relation


I expect that a user will write the following query instead:
> 
> ```sql
> SELECT hypertable_size(hypertable_name), hypertable_name FROM timescaledb_information.hypertables;
> ```

We can't prevent users from doing the other query so we still need to decide on what to do in that case, irrespective on whether the belief is that it is useful or not.

**Comment by erimatnor** [comment]
The feedback has been quite inconclusive so far, but trying to put this together I arrived at the current behavior:

* size functions return NULL when executed on non-hypertable objects --- Rationale: this was suggested by at least one person and seems consistent with the behavior of some PG size functions (although PG functions return  0 in some cases, e.g, indexes size on an index relation returns 0 since there are no indexes on an index). PG size functions seem to never return errors.
* Include size of root table of a hypertable; Rationale: the root table can have non-zero size objects, like indexes, even if it holds no data.
* Include an entry for the access node in detailed size output; Rationale: the access node has a hypertable that holds no data but can have non-zero size. In the future, the hypertable on the access node could potentially hold data or grow in size utilization for other reasons. Note that the node_name in the detailed output will be NULL for the access node since it has no user-given name.
